typed-racket-lib/typed-racket/rep/object-rep.rkt

```diff
@@ -6,43 +6,289 @@
+   (if new-const
+       (make-LExp* new-const new-terms)
+       ;; if const is #f then some term(s) became Empty
+       -empty-obj)]
+  [#:for-each (f) (for ([p (in-terms-vars (exp terms))]) (f p))]
+  [#:custom-constructor/contract
+   (-> exact-integer?
+       (hash/c Path? (and/c exact-integer?
+                            (not zero?)))
+       LExp?)
+   (intern-double-ref!
+    LExp-intern-table
+    terms const #:construct (make-LExp const terms))])
+
+(define LExp-intern-table (make-weak-hash))
+
+
+;; make-LExp* (provided as make-LExp)
+;;
+; IF the lexp (exp) contains only 1 variable,
+;;   and that variables its coefficient is 1,
+;    and the constant is 0
+; THEN that lone variable is returned
+; ELSE it returns the LExp
+;; NOTE 1: We do this so there is a 'canonical form'
+;; for linear expressions which are actually just
+;; the underlying object, e.g. insteaf of
+;; the LExp '0 + 1*x' we just want to return 'x'
+;; NOTE 2: we will also provide 'make-LExp'
+;; as 'raw-make-LExp' so that LeqProps can
+;; have both the lhs and rhs be LExps regardless
+;; if they are equivalent to a simpler object
+(define/cond-contract (make-LExp* const terms)
+  (-> exact-integer? hash? (or/c LExp? Path?))
+  (cond
+    [(and (eqv? 0 const)
+          (eqv? 1 (terms-count terms))
+          (for/fold ([res #f])
+                    ([(obj coeff) (in-terms terms)])
+            (and (eqv? 1 coeff) obj)))]
+    [else (make-LExp const terms)]))
+
+
+;; *****************************************************************************
+;; Operations for Linear Expressions
+
+
+;; constructor for LExps
+(define/cond-contract (-lexp . raw-terms)
+  (->* () () #:rest (listof (or/c exact-integer?
+                                  name-ref/c
+                                  Path?
+                                  (list/c exact-integer? (or/c name-ref/c Path?))))
+       LExp?)
+  (define-values (const terms)
+    (for/fold ([c 0] [ts (make-terms)])
+              ([term (in-list raw-terms)])
+      (match term
+        [(list (? exact-integer? coeff) (? Path? p))
+         (values c (terms-set ts p (+ coeff (terms-ref ts p))))]
+        [(list (? exact-integer? coeff) (? name-ref/c nm))
+         (let ([p (-id-path nm)])
+           (if (Empty? nm)
+               (values c ts)
+               (values c (terms-set ts p (+ coeff (terms-ref ts p))))))]
+        [(? exact-integer? new-const)
+         (values (+ new-const c) ts)]
+        [(LExp: c* ts*)
+         (values (+ c c*)
+                 (for/fold ([ts ts])
+                           ([(p coeff) (in-terms ts*)])
+                   (terms-set ts p (+ coeff (terms-ref ts p)))))]
+        [(? Object? p)
+         (values c (terms-set ts p (add1 (terms-ref ts p))))]
+        [(? name-ref/c var)
+         (define p (-id-path var))
+         (values c (terms-set ts p (add1 (terms-ref ts p))))])))
+  (make-LExp* const terms))
+
+
+;; LExp-add1
+(define/cond-contract (-lexp-add1 l)
```

Comment:
add1 is used for negation and expressing strict comparisons -- sub1 is useful when we recognize something like a strict comparison (e.g. while printing) and we want to remove the +1 to print it more cleanly w/ < or > instead of <= or >=